feat: add oidc support

Oniokey opened the pull request:
add support for oidc login.

close #508
close #554

我已确认该 PR 已自测通过，相关截图如下：
![image](https://github.com/user-attachments/assets/8cde6139-1622-454d-bd9e-d038c4c69d3e)
![image](https://github.com/user-attachments/assets/44d3d275-7715-4901-994e-815b463eca03)
![image](https://github.com/user-attachments/assets/77f234d2-0209-48a6-843d-5027713ed78b)

Oniokey committed:
feat: add the ui for configuring the third-party standard OAuth2.0/OIDC.

- update SystemSetting.js
- add setup ui
- add configuration

Oniokey committed:
feat: add the ui for "allow the OAuth 2.0 to login"

- update SystemSetting.js

Oniokey committed:
feat: add OAuth 2.0 web ui and its process functions

- update common.js
- update AuthLogin.js
- update config.js

Oniokey committed:
fix: missing "Userinfo" endpoint configuration entry, used by OAuth clients to request user information from the IdP.

- update config.js
- update SystemSetting.js

Oniokey committed:
feat: updated the icons for Lark and OIDC to match the style of the icons for WeChat, EMail, GitHub.

- update lark.svg
- new oidc.svg

Oniokey committed:
refactor: Changing OAuth 2.0 to OIDC

Oniokey committed:
feat: add OIDC login method

Oniokey committed:
feat: Add support for OIDC login to the backend

LeoQuote commented:
特别期待的功能! 

建议统一术语为标准术语: https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1

client id, client secret, 不使用 appID, appSecret 等非标准术语.

建议支持 oidc 的自动发现 https://auth0.com/docs/get-started/applications/configure-applications-with-oidc-discovery 

userinfo_endpoint, token_endpoint, authorization_endpoint 改为高级选项, 用于兼容不支持自动发现的提供商, 这样方便配置.

这些术语个人建议不要翻译, 因为会配置的自然也懂, 不会配置的翻译了也没用, 和供应商的术语还有可能对不上.

Oniokey commented:
> client id, client secret, 不使用 appID, appSecret 等非标准术语.

遵循现有的标准是很好的建议，非常乐意修改为标准术语

> 建议支持 oidc 的自动发现，将各类端点改为高级选项

很好的想法，我的实现思路是提供一个是否使用自定义端点的选项，默认不启用，使用  well-known endpoint 自动发现，手动启用后，切换为现有模式。并且在前端完成自动发现：通过 GET 请求 well-known endpoint 并解析返回值中的各个端点，之后通过现有的方式保存到后端

Oniokey committed:
fix: Change the AppId and AppSecret on the Web UI to the standard usage: ClientId, ClientSecret.

Oniokey committed:
feat: Support quick configuration of OIDC through Well-Known Discovery Endpoint

LeoQuote commented:
我个人认为不应该将自动发现的 endpoint 保存至后端, 实践中虽然很少有接口更改的情况, 但该机制的初衷就是希望通过自动发现来自动配置, 不需要用户的操作, 如果提供商有接口地址更新, 有可能会导致这边的配置无法使用, 造成一些不符合直觉的现象.

比较建议实例化时实时获取最新的配置, 这样即使说提供商有接口地址更改, 一般重启应用即可解决问题, 而如果将 endpoint 保存至后端, 还需要手动清除相关的高级配置才可真正更新.

LeoQuote commented:
还有就是可以考虑使用 https://github.com/coreos/go-oidc 这类的第三方包进行实现, 这种包直接提供了自动发现的相关机制, 无需重复实现.

Oniokey commented:
> 我个人认为不应该将自动发现的 endpoint 保存至后端, 实践中虽然很少有接口更改的情况, 但该机制的初衷就是希望通过自动发现来自动配置, 不需要用户的操作, 如果提供商有接口地址更新, 有可能会导致这边的配置无法使用, 造成一些不符合直觉的现象.
> 
> 比较建议实例化时实时获取最新的配置, 这样即使说提供商有接口地址更改, 一般重启应用即可解决问题, 而如果将 endpoint 保存至后端, 还需要手动清除相关的高级配置才可真正更新.

存储 自动发现端点 的作用是在配置结束后再次进入到配置页面可以得到之前配置的内容是什么，仅为存储用途，对于后端并无功能性作用，自动发现的过程将在前端完成。

Oniokey commented:
> 还有就是可以考虑使用 https://github.com/coreos/go-oidc 这类的第三方包进行实现, 这种包直接提供了自动发现的相关机制, 无需重复实现.

感谢提醒，目前后端部分已经完全按照 OIDC 标准实现了，OIDC 标准并不复杂，实现起来并不困难，也鲜有出现 Bug 的机会，所以这里不再修改了，如果后续此版 OIDC 实现出现 Bug 会再次 pr 一个使用此库的版本。

Oniokey committed:
feat: Standardize terminology, add well-known configuration

- Change the AppId and AppSecret on the Server End to the standard usage: ClientId, ClientSecret.
- add Well-Known configuration to store in database, no actual use in server end but store and display in web ui only

Oniokey commented:
pr 已更新

songquanpeng commented:
Thx~

wzxjohn commented:
这个实现问题有点多

1. Token 请求未遵循规范 #2105 
2. 未正确处理异常响应，不论是否鉴权成功均会创建用户
3. 设置若不填写选填的 Well Known 会前端报错，无法保存

Oniokey commented:
> 这个实现问题有点多
> 
> 1. Token 请求未遵循规范 [OIDC功能token接口请求体不正确 #2105](https://github.com/songquanpeng/one-api/issues/2105)
> 2. 未正确处理异常响应，不论是否鉴权成功均会创建用户
> 3. 设置若不填写选填的 Well Known 会前端报错，无法保存

确实是存在很多问题的
1. 这个 pr 主要是参考了 Casdoor 关于 OIDC 的流程实现的，部分操作可能并不符合 RFC 规范，欢迎 pr
> 不过确实很多 IdP 在实现上不尽相同，比如最近就发现 Casdoor 和 Logto 在 SAML 和 OIDC 上的部分实现并不相同
3. 后面两个 Bug 会找时间修复，感谢反馈

noiceman commented:
为什么我用官方docker镜像没有这个功能，是要源码部署吗？
